### PR TITLE
bazel: use CC libraries from nix

### PR DESCRIPTION
**Pull request by malt3**
### Context

This is a refactoring that simplifies dependency management and allows us to run integration tests with hermetic dependencies (in Bazel).
Before this PR, binaries that we ship in the Constellation OS images (bootstrapper, disk-mapper) link against Fedora libraries (libcryptsetup).
Doing this correctly and hermetically was challenging, as it required us to lock Fedora rpms using bazeldnf.
This also prevents us from testing Go binaries that use cgo effectively, since any Go binaries produced as a result of linking against libraries from Fedora are unusable on any platform but Fedora.

The new approach is to import libraries from nix / nixpkgs and ship the required runtime dependencies where needed.

### Proposed change(s)

- bazel: use CC libraries from nix

### Additional info

- More testing is needed before this is ready
- [image build pipeline (debug)](https://github.com/edgelesssys/constellation/actions/runs/7048025167): `ref/feat-bazel-use-cc-libraries-from-nix/stream/debug/v2.14.0-pre.0.20231130153434-2992ec0d0aa9`
- [image build pipeline (nightly)](https://github.com/edgelesssys/constellation/actions/runs/7048031681): `ref/feat-bazel-use-cc-libraries-from-nix/stream/nightly/v2.14.0-pre.0.20231130153434-2992ec0d0aa9`
- [test run for uploading the libvirt base image](https://github.com/edgelesssys/constellation/actions/runs/7017027391)
- [e2e azure](https://github.com/edgelesssys/constellation/actions/runs/7018912901), [e2e gcp](https://github.com/edgelesssys/constellation/actions/runs/7019148660), [e2e aws](https://github.com/edgelesssys/constellation/actions/runs/7019159220)
- Part of [AB#3067](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3067)

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: cli/internal/libvirt/README.md
Comment (daniel-weisse) [inline cli/internal/libvirt/README.md:1]: We should probably move this README.
Its currently rather confusing to have build instructions for a container image in a directory that contains none of the files that are required to built it.

**Review by daniel-weisse — APPROVED** [review]

**Review by 3u13r — APPROVED** [review]
LGTM

File: nix/container/libvirtd_base.nix [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    systemd-oom:x:997:997:systemd Userspace OOM Killer:/:/usr/sbin/nologin
+    systemd-resolve:x:193:193:systemd Resolver:/:/usr/sbin/nologin
+    polkitd:x:996:996:User for polkitd:/:/sbin/nologin
+    rtkit:x:172:172:RealtimeKit:/proc:/sbin/nologin
+    gluster:x:995:995:GlusterFS daemons:/run/gluster:/sbin/nologin
+    dnsmasq:x:994:994:Dnsmasq DHCP and DNS server:/var/lib/dnsmasq:/usr/sbin/nologin
+    rpc:x:32:32:Rpcbind Daemon:/var/lib/rpcbind:/sbin/nologin
+    rpcuser:x:29:29:RPC Service User:/var/lib/nfs:/sbin/nologin
+    qemu:x:107:107:qemu user:/:/sbin/nologin
+    pipewire:x:993:992:PipeWire System Daemon:/var/run/pipewire:/sbin/nologin
```
Comment (burgerdev) [inline nix/container/libvirtd_base.nix:34]: I see that this is only moved, but maybe we can clean this up eventually? :D

**Review by burgerdev — COMMENTED** [review]
This looks like a huge improvement, great job!

File: nix/container/libvirtd_base.nix [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      mkdir -p /var/lock
+      mkdir -p /var/lib/libvirt/boot
+      mkdir -p /var/lib/libvirt/dnsmasq
+      mkdir -p /var/lib/libvirt/filesystems
+      mkdir -p /var/lib/libvirt/images
+      mkdir -p /var/lib/libvirt/libxl
+      mkdir -p /var/lib/libvirt/lxc
+      mkdir -p /var/lib/libvirt/network
+      mkdir -p /var/lib/libvirt/qemu
+      mkdir -p /var/lib/libvirt/swtpm
```
[the remark this answers is not in the document]
Comment (malt3) [reply nix/container/libvirtd_base.nix:130]: Good idea. I'll give it a try.

File: .github/workflows/build-libvirt-container.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build container
+        run: |
+          nix build .#libvirtd_base
+          gunzip < result > libvirtd_base.tar
+          crane push libvirtd_base.tar ghcr.io/edgelesssys/constellation/libvirtd-base
```
[the remark this answers is not in the document]
Comment (malt3) [reply .github/workflows/build-libvirt-container.yml:43]: Yes this works fine in practice. The container is pinned via hash and renovate should be capable of updating the pinned container in `bazel/toolchains/container_images.bzl`.

File: nix/cc/nixpkgs_cc_libraries.bzl [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    visibility = ["//visibility:public"],
+)
+"""
+
+CRYPTSETUP_BUILD = """\
+filegroup(
+    name = "include",
+    srcs = glob(["include/**/*.h"]),
+    visibility = ["//visibility:public"],
+)
```
Comment (burgerdev) [inline nix/cc/nixpkgs_cc_libraries.bzl:74]: Could we reuse the built-in targets https://github.com/tweag/rules_nixpkgs/blob/21c4ea481021cb51a6e5d0969b2cee03dba5a637/core/BUILD.bazel.tpl#L14?

File: bazel/patchelf/patchelf.bzl [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        ],
+        command = "\"$1\" --set-rpath \"$(cat \"$2\")\" --output \"$3\" \"$4\"",
+        progress_message = "Patching ELF binary " + ctx.file.src.basename,
+    )
+    return DefaultInfo(
+        files = depset([output]),
+        executable = output,
+    )
+
+patchelf = rule(
```
[the remark this answers is not in the document]
Comment (malt3) [reply bazel/patchelf/patchelf.bzl:24]: Added: https://github.com/edgelesssys/constellation/blob/5009cbabac3bf244ca83ad170c3685a3daf38cc1/bazel/patchelf/patchelf.bzl#L38-L49

File: bazel/toolchains/container_images.bzl [hunk 6]
```diff
@@ -14,3 +14,8 @@ def containter_image_deps():
             "linux/arm64",
         ],
     )
+    oci_pull(
+        name = "libvirtd_base",
+        digest = "sha256:08d0643e47365d9a928e9b08845e932834da9e445b4ee96fd0f1f721eebcea07",
+        image = "ghcr.io/edgelesssys/constellation/libvirtd-base",
```
[the remark this answers is not in the document]
Comment (malt3) [reply bazel/toolchains/container_images.bzl:20]: I updated the top-level comment to be more correct. 

The current approach is this:

- CI builds and pushes libvirt-base using only nix
- Renovated updates this oci_pull rule
- In Bazel, we refer to the finished libvirt-base image and can push it into the developers container registry during `bazel run //:devbuild`

File: cli/internal/libvirt/README.md [hunk 7]
```diff
@@ -11,14 +11,19 @@ Connecting to the libvirt daemon running in the container and manage the deploym
...
-Build the image:
+Update the base image:
+
+```shell
+nix build .#libvirtd_base
+cat result | gunzip > libvirtd_base.tar
+crane push libvirtd_base.tar ghcr.io/edgelesssys/constellation/libvirtd-base
+```
+
+Push to your own registry:
```
[the remark this answers is not in the document]
Comment (malt3) [reply cli/internal/libvirt/README.md]: I updated the wording slightly. Does this make sense with the explanation here? https://github.com/edgelesssys/constellation/pull/2641#discussion_r1410716556

If not, I should definitely update improve the README further.

File: nix/cc/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+)
+
+alias(
+    name = "libvirt_rpath",
+    actual = select({
+        ":aarch64-linux": "@libvirt_aarch64-linux//:libvirt",
+        ":x86_64-linux": "@libvirt_x86_64-linux//:libvirt",
+    }),
+    visibility = ["//visibility:public"],
+)
```
[the remark this answers is not in the document]
Comment (malt3) [reply nix/cc/BUILD.bazel]: Removed

File: nix/container/libvirtd_base.nix (hunk 1 above)
Comment (malt3) [reply nix/container/libvirtd_base.nix]: Cleaned up unneded users / groups:

https://github.com/edgelesssys/constellation/blob/b424a15cb7f4ad8b63ad677bfea72900a6dc20bf/nix/container/libvirtd_base.nix#L6-L52

File: nix/container/libvirtd_base.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (malt3) [reply nix/container/libvirtd_base.nix]: Done :)

https://github.com/edgelesssys/constellation/blob/b424a15cb7f4ad8b63ad677bfea72900a6dc20bf/nix/container/libvirtd_base.nix#L117-L138

File: nix/cc/nixpkgs_cc_libraries.bzl (hunk 4 above)
Comment (malt3) [reply nix/cc/nixpkgs_cc_libraries.bzl:73]: I think we need to customize our `BUILD` file to incorporate the `cc_library` and `export_files` rules.
nixpkgs will also only provide the built-in targets if `build_file_content` is not set:

https://github.com/tweag/rules_nixpkgs/blob/21c4ea481021cb51a6e5d0969b2cee03dba5a637/core/nixpkgs.bzl#L474-L482

File: nix/cc/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        ":x86_64-linux": "@org_openssl_x86_64-linux//:org_openssl",
+    }),
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "cryptsetup",
+    actual = select({
+        ":aarch64-linux": "@cryptsetup_aarch64-linux//:cryptsetup",
+        ":x86_64-linux": "@cryptsetup_x86_64-linux//:cryptsetup",
```
[the remark this answers is not in the document]
Comment (burgerdev) [reply nix/cc/BUILD.bazel]: My mental model of the Bazel build is a bit sketchy, I don't know whether my idea is viable at all. The last adjustments to this file lgtm, let's leave it at that for now.

File: cli/internal/libvirt/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (burgerdev) [reply cli/internal/libvirt/README.md]: Yep, looks good now.

File: nix/cc/nixpkgs_cc_libraries.bzl (hunk 4 above)
Comment (burgerdev) [reply nix/cc/nixpkgs_cc_libraries.bzl:73]: Understood

**Review by burgerdev — APPROVED** [review]
lgtm

**Pull request merged**